packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +157,263 @@ if (process.env.NODE_ENV === 'development') {
+   * - Set `visibleCount` to the `maxGuess`.
+   * - Repeat the following by calling `guessVisibleCount()` (keep re-rendering but not painting):
+   *   - Each time the container overflows, new max guess is the average of the two guesses.
+   *   - Each time the container does not overflow, new min guess is the average of the two guesses.
+   * - Stop when the average of the two guesses is the min guess itself. i.e. no more averaging possible.
+   * - The min guess is then the correct `visibleCount`.
+   */
+  const guessVisibleCount = React.useCallback(() => {
+    // If already stabilized, already guessing, or in unit test, do not guess.
+    if (isStabilized || isGuessing.current || isUnitTest) {
+      return;
+    }
+
+    try {
+      isGuessing.current = true;
+
+      // We need to wait for the ref to be attached so that we can measure available and required sizes.
+      if (containerRef.current == null) {
+        return;
+      }
+
+      // If there are no items, stabilize immediately.
+      if (itemsCount === 0) {
+        dispatch({ type: 'stabilize' });
+        return;
+      }
+
+      const dimension = orientation === 'horizontal' ? 'Width' : 'Height';
+      const availableSize = containerRef.current[`offset${dimension}`];
+      const requiredSize = containerRef.current[`scroll${dimension}`];
+
+      const isOverflowing = availableSize < requiredSize;
+
+      if (
+        // We have already found the correct visibleCount
+        (visibleCount === itemsCount && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (maxGuess - minGuess === 1 && visibleCount === minGuess)
+      ) {
+        dispatch({ type: 'stabilize' });
+        return;
+      }
+
+      // Before the main logic, the max guess MUST be >= the correct visibleCount for the algorithm to work.
+      // If not:
+      // - double the max guess and visibleCount: since we need to overflow.
+      // - set min guess to current visibleCount: since underflow means correct visibleCount >= current visibleCount.
+      if (maxGuess === visibleCount && !isOverflowing) {
+        dispatch({ type: 'shiftGuessRangeForward' });
+        return;
+      }
+
+      let newMinGuess = minGuess;
+      let newMaxGuess = maxGuess;
+
+      if (isOverflowing) {
+        // overflowing = we guessed too high. So, new max guess = half the current guess
+        newMaxGuess = visibleCount;
+      } else {
+        // not overflowing = maybe we guessed too low? So, new min guess = half of current guess
+        newMinGuess = visibleCount;
+      }
+
+      // Next guess is always the middle of the new guess range
+      const newVisibleCount = Math.floor((minGuess + maxGuess) / 2);
+
+      // If newVisibleCount is 0 (impossible state as min is 1), end guessing
+      if (newVisibleCount === 0) {
+        dispatch({ type: 'stabilize' });
+        return;
```

Comment:
Why do we need to handle this case again randomly here? Shouldn't it already be handled by one of the `if` checks earlier?